### ISSUE TITLE
[CN] Hard to figure out that `extract` needs `u64` argument for array-typed blocks

### ISSUE THREAD

**Issue by septract** [rems-project/cerberus#256]
When writing `extract` for array-typed blocks, the operator takes a `u64` value. This is hard to figure out, especially since the C-level index is an `int` / `i32`. 

Example: 
```
void
array_4()
{
    int arr[1];
    /*@ extract Block<int>, 0u64; @*/ // <-- non-obvious 
    arr[0] = 1;
}
```

See also the discussion with @cp526 on MM [here](https://mattermost.galois.com/verse-ta1-ta2/pl/g5e7xy5knib5f85euqx4iq7isc). Quoting the relevant part: 

> @cp526: "With CN as-is the `extract` needs a `u64` index, because a `Block<int[1]>` (or any other array-typed block) decomposes into an iterated resource indexed by `u64`. [...] That's not great UI-wise (having to match that so-far-invisible u64 index type), so I have to see what we do about that. Perhaps if we adopt your other suggestion (from a while back) of having `extract` additionally take the base pointer of the resource the extract should be applied to (which we should just do anyway), CN could provide some useful output about why nothing was extracted.

**Comment by septract** [comment]
You need the same `u64` type for asserting properties of array offsets, eg the following (which is example 00090 in the `c-testsuite` dataset) 
```
int a[3] = {0, 1, 2};

int
main()
/*@ accesses a; @*/
/*@ requires 
			a[0u64] == 0i32; 
			a[1u64] == 1i32; 
			a[2u64] == 2i32; 
			@*/
/*@ ensures return == 0i32; @*/
{
	/*@ extract Owned<int>, 0u64; @*/
	/*@ extract Owned<int>, 1u64; @*/
	/*@ extract Owned<int>, 2u64; @*/
	if (a[0] != 0)
		return 1;
	if (a[1] != 1)
		return 2;
	if (a[2] != 2)
		return 3;
	
	return 0;
}
```

This is also not very intuitive.

**Comment by peterohanley** [comment]
If your `each` statements do not have `u64` indices, they will also not be used to satisfy resources.

**Comment by septract** [comment]
This issue showed up again in an example from BAE / @lwli11

**Comment by jprider63** [comment]
I've started looking into this. I'm somewhat confused by what's going on. Examples like the following fail as described:
```
void
array_4()
{
    int arr[1];
    /*@ extract Block<int>, 0i32; @*/
    arr[0] = 1;
}
```

However the following [example from the tutorial](https://rems-project.github.io/cn-tutorial/exercises/array_load.c) passes even though an `i32` is used as the index in `each` and `extract`:
```
int read (int *p, int n, int i)
/*@ requires take a1 = each(i32 j; 0i32 <= j && j < n) {
                          Owned<int>(array_shift<int>(p,j)) };
             0i32 <= i && i < n;
    ensures take a2 = each(i32 j; 0i32 <= j && j < n) {
                            Owned<int>(array_shift<int>(p,j)) }; @*/
{
  /*@ extract Owned<int>, i; @*/
  return p[i];
}
```

What's the difference here that causes the second example to pass while the first one fails?

**Comment by septract** [comment]
@jprider63 My interpretation is that CN will allow you to specify in a pre-condition that an array is `i32` indexed, but the CN interpretation of the C semantics will not allow you to _construct_ such an array at runtime. 

So in `array_4()` you are trying to access an array which was allocated on the stack, which is `u64` indexed. But in `read` you require the existence of an `i32`-indexed array. 

I think this is the crux of the confusion in fact. I think you definitely should not be able to specify in pre-conditions memory states that cannot be constructed at runtime.

**Comment by cp526** [comment]
> @jprider63 My interpretation is that CN will allow you to specify in a pre-condition that an array is `i32` indexed, but the CN interpretation of the C semantics will not allow you to _construct_ such an array at runtime.

Memory ownership that comes from proper C arrays (e.g. `int[5]`) will indeed not be i32-indexed, but other iterated resources might have different index types, e.g. memory obtained from some allocator could perhaps be specified using a different index type.

> 
> So in `array_4()` you are trying to access an array which was allocated on the stack, which is `u64` indexed. But in `read` you require the existence of an `i32`-indexed array.
> 
> I think this is the crux of the confusion in fact. I think you definitely should not be able to specify in pre-conditions memory states that cannot be constructed at runtime.

I think the real source of the problem is that the user has no good way of knowing that a C array always turns into a u64-indexed `each`. But we could decide to tighten the wellformedness rules and only ever allow `u64`-indexed arrays. In that case one could make `extract` statements with indices other than `u64` index type an error and avoid the confusing situation above. 

I don't know if there's ever a very good reason for choosing a non-u64 index type. Though in principle, one could imagine in the future supporting index types that are not even integer-typed, such as an each where the index is drawn from a set of values (as in the Viper paper CN's iterated resources were originally inspired by).

**Comment by septract** [comment]
Yes, I agree - the issue for the user is they can't discover they need the `u64` and the error message doesn't help.

Another option would be to silently up-cast `i32` to `u64` in situations like the above, which is kind of what C is doing when it's passed an `int` as an array argument. (I'm not sure what happens if you pass a negative `int` though...)

**Comment by jprider63** [comment]
Thanks, that's helpful! For now, it sounds like `cn` should throw a type error if the user uses the wrong index type with an error message that clearly indicates what the correct index type should be. I'll work on this unless there's another solution you prefer.

**Comment by septract** [comment]
You might also get it to issue a warning when something other than u64 is used to index an array, since that seems to be nearly always the wrong thing in the current design.

**Comment by peterohanley** [comment]
If `cn` is being used to specify code for a 32-bit machine would the correct index type become `u32`?

**Comment by cp526** [comment]
> Thanks, that's helpful! For now, it sounds like `cn` should throw a type error if the user uses the wrong index type with an error message that clearly indicates what the correct index type should be. I'll work on this unless there's another solution you prefer.

I think for now that means: iterated resources and extract both have to use `u64` for the index type, which potentially requires changing some examples we've got.

In principle one might want CN to somehow reject `extract` statements whose index does not match the `each`, but that's not something we can do at the moment, because an `extract` does not uniquely identify a paired resource.

**Comment by cp526** [comment]
> If `cn` is being used to specify code for a 32-bit machine would the correct index type become `u32`?

That's something we should maybe consider. In `baseTypes.ml` there is logic for translating each C type to the corresponding CN base type. In general this is machine dependent (e.g. in the case of an `int`), so that translation is set up to query Cerberus's record of the implementation choices (which Cerberus is parametric in), but I'm not sure what CN should do for array indices.

**Comment by peterohanley** [comment]
I think it should stay u64: C converts the index type anyway, and if it's u32 on a 32-bit platform code will have to be specified for a certain platform. Pointer width is a separate issue.
